### DISCUSSION TITLE
Multiplexing Audio with Raspotify, Shairport-Sync, and PipeWire on Pi OS Lite

### DISCUSSION THREAD

**Discussion by giggywithit** [dtcooper/raspotify#691] [Q&A]
Hello community,

I'm setting up a Raspberry Pi running Pi OS Lite to serve as an audio hub, capable of streaming music via Spotify Connect (using Raspotify) and AirPlay (using Shairport-Sync) simultaneously. My goal is to multiplex audio outputs so that I can play music from both services at the same time through the same audio device. Here are my current challenges and questions:

PipeWire on Pi OS Lite:
Pi OS Lite doesn't come with PipeWire pre-installed. What's the best way to install and configure PipeWire for audio multiplexing on this minimal OS? Are there specific settings or configurations I should be aware of?
Multiplexing Audio:
Has anyone successfully managed to multiplex audio from both Raspotify and Shairport-Sync using PipeWire? If so, could you share your setup or configuration tips?
What are the potential pitfalls of running PipeWire alongside these services for audio multiplexing?
ALSA and dmix:
I've tried using ALSA's dmix plugin but haven't had any success. Are there known issues with dmix on Pi OS Lite, or is there a better approach for mixing audio outputs from multiple sources?
Alternative Solutions:
If PipeWire isn't the best solution for this use case, what alternatives exist for audio multiplexing on a Raspberry Pi with Pi OS Lite?
Specific Questions about Raspotify and Shairport-Sync:
Do I need to modify how these services interact with the audio system to work with PipeWire, or can they continue to use ALSA while PipeWire manages the multiplexing?

Here's my current Raspotify configuration:

LIBRESPOT_QUIET=
LIBRESPOT_ALSA_MIXER_DEVICE="hw:0"
LIBRESPOT_MIXER="alsa"
LIBRESPOT_ALSA_MIXER_CONTROL="Digital"
LIBRESPOT_NAME="COX-GhostAmp"
LIBRESPOT_BITRATE="320"
LIBRESPOT_INITIAL_VOLUME="20"
LIBRESPOT_VOLUME_CTRL="cubic"
LIBRESPOT_AUTOPLAY="on"
LIBRESPOT_DISABLE_AUDIO_CACHE=
LIBRESPOT_DISABLE_CREDENTIAL_CACHE=
LIBRESPOT_ENABLE_VOLUME_NORMALISATION=
TMPDIR=/tmp

Any insights, experiences, or configurations you could share would be greatly appreciated. Thanks in advance for your help!

**Comment by kimtore** [comment]
Hi @giggywithit,

I've set up a similar setup on vanilla Debian 12, with Pipewire running with a separate user, started through the global systemd. I use version 1.2.x from the debian-backports repository.

It is easiest if applications use the Pulseaudio protocol to speak with Pipewire, due to permissions and dbus architecture.

This setup requires quite a bit of systemd plumbing, user setup, config tweaking. I took some inspiration from this issue from the Pipewire bug tracker, perhaps you could apply the same to get it working. https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/2196

Good luck, feel free to ask for more details.

**Comment by kimtore** [comment]
As an example I might mention that I run Kodi, Raspotify, MPD, Moonlight, and a RTP audio receiver through Pipewire on the same Raspberry Pi. All of them connect effortlessly over Pulseaudio, to be honest it feels quite smooth and I will definitely recommend Pipewire as the multiplexer.

**Comment by giggywithit** [comment]
Thanks for the input...
My only real desire for pipewire is for multiplex of two sources, as I have had no success using dmix, but pipewire doesn't seem to really have any intention of being used headless or systemwide.. Perhaps I am under the wrong impression that Pulse is looking to be superceded in any way. Maybe pulse is here to stay
Thanks

**Reply by kimtore** [reply]
Well, Pipewire is definitely usable this way, even though that's not their intent.

In my personal experience, Pipewire is powerful and flexible and has done everything I wanted it to, both in desktop and pro audio applications. It provides ALSA, Pulseaudio and JACK without any fussing around.

I was thinking of writing a guide since there are limited resources online to figure this out, but it still feels a bit hackish so perhaps a Debian package is better.
